spire: Fix unreliable test

joestringer opened the pull request:
Since the introduction of private FIPS fields in core library RSA
key fields, the Certificate.PrivateKey fields can differ when generated
from the same key, as observed by `reflect.DeepEqual`. If you use the native
type's `.Equal()` function, it reports the fields are the same.

This difference only appears when running this test frequently or many
times in a row, such as by using `go test -count=1000 -failfast ...`

Expand out the checks for the Certificate's child fields to validate
the Certificate, PrivateKey and Leaf independently. While we're at it,
switch to `assert.Equal` so that there's a higher chance that the output
can be interpreted by developers when the test fails.

Fixes: https://github.com/cilium/cilium/issues/40130

joestringer committed:
spire: Fix unreliable test

Since the introduction of private FIPS fields in core library RSA
key fields, the Certificate.PrivateKey fields can differ when generated
from the same key, as observed by 'reflect.DeepEqual'. If you use the native
type's '.Equal()' function, it reports the fields are the same.

This difference only appears when running this test frequently or many
times in a row, such as by using 'go test -count=1000 -failfast ...'

Expand out the checks for the Certificate's child fields to validate
the Certificate, PrivateKey and Leaf independently. While we're at it,
switch to 'assert.Equal' so that there's a higher chance that the output
can be interpreted by developers when the test fails.

Signed-off-by: Joe Stringer <joe@cilium.io>

mhofstetter commented on pkg/auth/spire/certificate_provider_test.go:
```diff
@@ -389,8 +390,24 @@ func TestSpireDelegateClient_GetCertificateForIdentity(t *testing.T) {
+				wantKey := tt.want.PrivateKey.(*rsa.PrivateKey)
+				gotKey := got.PrivateKey.(*rsa.PrivateKey)
+				if !wantKey.Equal(gotKey) {
+					if !assert.Equal(t, tt.want.PrivateKey, got.PrivateKey) {
```
you can add a message to the assert functions instead of checking their result and emitting an additional error.

```
assert.Equal(t, tt.want.PrivateKey, got.PrivateKey, "SpireDelegateClient.GetCertificateForIdentity().PrivateKey = %v, want %v", got.PrivateKey, tt.want.PrivateKey)
```

joestringer replied:
Yeah that can be helpful in general. For many of these cases, `assert.Equal` will already emit the relevant information for us in a standardized form. Where this PR uses `assert.Equal`, I'll just drop the message entirely.

mhofstetter commented on pkg/auth/spire/certificate_provider_test.go:
```diff
@@ -389,8 +390,24 @@ func TestSpireDelegateClient_GetCertificateForIdentity(t *testing.T) {
+					t.Errorf("SpireDelegateClient.GetCertificateForIdentity().Leaf = %v, want %v", got.Leaf, tt.want.Leaf)
+				}
+			} else {
+				if !assert.Equal(t, tt.want, got) {
```
-> `assert.Nil(t, got, "SpireDelegateClient.GetCertificateForIdentity() = %v, should be nil")` ?

mhofstetter commented on pkg/auth/spire/certificate_provider_test.go:
```diff
@@ -389,8 +390,24 @@ func TestSpireDelegateClient_GetCertificateForIdentity(t *testing.T) {
+				}
+				wantKey := tt.want.PrivateKey.(*rsa.PrivateKey)
+				gotKey := got.PrivateKey.(*rsa.PrivateKey)
+				if !wantKey.Equal(gotKey) {
```
is it really necessary to have this additional equality check on the type itself - even though we have the following equals assertion?

IMO if one of them isn't reliable we should simply not use it for assertion at all.

joestringer replied:
Yeah it's necessary unfortunately. Upstream core libraries have private fields that differ, but they provide an `.Equal()` function to determine equality. That's the underlying reason for the unreliability of the test.

Similar to your other feedback on this PR, we can tidy up the logic here to simplify the test logic. It looks like there's a few test files that have similar poor patterns for use of `testify.assert`, will take note that could do with some follow up.

mhofstetter commented on pkg/auth/spire/certificate_provider_test.go:
```diff
@@ -389,8 +390,24 @@ func TestSpireDelegateClient_GetCertificateForIdentity(t *testing.T) {
 				t.Errorf("SpireDelegateClient.GetCertificateForIdentity() error = %v, wantErr %v", err, tt.wantErr)
```
maybe worth to refactor this to use testify too - while we're at it.

```
assert.Equal(t, tt.wantErr, err != nil, "SpireDelegateClient.GetCertificateForIdentity() error = %v, wantErr %v", err, tt.wantErr)

```

joestringer replied:
Yeah makes sense. I'll apply similar changes to the rest of the file.

mhofstetter reviewed (commented):
Thanks Joe! I left some comments that might help to make the test more readable.

edit: maybe it would be worth to add a small comment in the code why we're comparing like this.

joestringer committed:
spire: Convert certificate unit tests to testify

Use testify/assert to assert the results of these unit tests, for
simpler syntax and more consistent output that reports more useful
information on failure.

Signed-off-by: Joe Stringer <joe@cilium.io>

mhofstetter reviewed (approved):
Thanks!
